zkevm-circuits/src/evm_circuit/util/memory_gadget.rs

```diff
@@ -686,3 +686,154 @@ impl<F: Field, const MAX_BYTES: usize, const ADDR_SIZE_IN_BYTES: usize>
+                .map(|e| buffer_reader_gadget.read_flag(e))
+                .collect::<Vec<Expression<F>>>();
+
+            // test byte API
+            for (byte_expr, buffer_reader_gadget_byte_expr) in bytes_expr.iter().zip(buffer_reader_gadget_bytes_expr.iter()) {
+                cb.require_equal(
+                    "every byte equal",
+                    byte_expr.expr(),
+                    buffer_reader_gadget_byte_expr.expr(),
+                );
+            }
+
+            // test read_flag API
+            cb.require_equal(
+                "selector length equal",
+                sum::expr(buffer_reader_gadget_seletor),
+                addr_end.expr() - addr_start.expr(),
+            );
+
+            BufferReaderGadgetTestContainer { 
+                buffer_reader_gadget, 
+                addr_start,
+                addr_end, 
+                bytes,
+            }
+        }
+
+        fn assign_gadget_container(
+            &self,
+            witnesses: &[Word],
+            region: &mut CachedRegion<'_, '_, F>,
+        ) -> Result<(), Error> {
+            let offset = 0;
+            let addr_start= u64::from_le_bytes(witnesses[0].to_le_bytes()[..8].try_into().unwrap());
+            let addr_end= u64::from_le_bytes(witnesses[1].to_le_bytes()[..8].try_into().unwrap());
+            let mut input_bytes: Vec<u8> = Vec::new();
+            input_bytes.extend_from_slice(&witnesses[2].to_le_bytes());
+            self.addr_start.assign(region, offset, Some(addr_start.to_le_bytes()))?;
+            self.addr_end.assign(region, offset, Some(addr_end.to_le_bytes()))?;
+            self.bytes
+                .iter()
+                .zip(input_bytes.iter())
+                .map(|(byte, input_byte)| 
+                    byte.assign(region, offset, Value::known(F::from(*input_byte as u64)))
+                )
+                .collect::<Result<Vec<_>, _>>()?;
+            self.buffer_reader_gadget.assign(region, offset, addr_start, addr_end, &input_bytes[..])?;
+
+            Ok(())
+        }
+    }
+    
+    #[test]
+    fn test_buffer_reader_gadget_completness(){
+        // buffer len = data len
+        try_test!(
+            BufferReaderGadgetTestContainer<Fr, 32, 10>,
+            vec![
+                Word::from(0), 
+                Word::from(2), 
+                Word::from(256),
+            ],
+            true,
+        );
+
+        // buffer len > data len
+        try_test!(
+            BufferReaderGadgetTestContainer<Fr, 32, 10>,
+            vec![
+                Word::from(0), 
+                Word::from(31),
+                Word::from(255),
+            ],
+            true,
+        );
+    }
+
+    #[test]
+    fn test_buffer_reader_gadget_soundness(){
+        // buffer len < data len
+        try_test!(
+            BufferReaderGadgetTestContainer<Fr, 32, 10>,
+            vec![
+                Word::from(0),
+                Word::from(1), 
+                Word::from(1024),
```

Comment:
nit: suggest to change the 3rd input to 256.

The data size in your first two cases are 2 bytes (256) and 1 byte (255) and this one is 2 bytes (1024). If you want to represent 2 bytes data, it's better to use either 256 or 1024 (just one fixed value). If so, reviewers could understand what you want to test easily. Different valus could represent different things. So, I would suggest to change this value to 256 (2 bytes data and the len is larger than the given buffer size) and change 255 to 256 in above case either (bcs your buffer is 31 bytes, way to larger than 2 bytes data)